### DISCUSSION TITLE
How to find chunks similar to EntityRuler

### DISCUSSION THREAD

**Discussion by kaykhancheckpoint** [explosion/spaCy#5924] [Help: Coding & Implementations]
Hi

Im wondering how the EntityRuler correctly splits nouns/phrases like "OpTic Gaming" & "Call of Duty" if i use the general noun_chunker it does not find this.

**code**:
```
nlp = spacy.load("en_core_web_sm")
doc = nlp("The company was also one of OpTic Gaming's main sponsors during the legendary organization's run to their first Call of Duty Championship back in 2017")
for chunk in doc.noun_chunks:
    print(chunk)
```
**output**:
```
The company
OpTic Gaming's main sponsors
their first Call
Duty Championship

```

What i would like is something like
```
...
OpTic Gaming
Call of Duty
...
```

**Comment by svlandeg** [comment]
The [noun chunker](https://spacy.io/usage/linguistic-features#noun-chunks) tries to identify base noun phrases by looking at the dependency parse. In your example sentence though, the dependency parsing is not able to parse "Call of Duty" correctly, resulting in wrong dependency relations and ultimately a chopped up noun phrase ("their first Call" and "Duty Championship"). Ideally, this should have been "their first Call of Duty Championship".

The focus of the noun chunker is on broad noun phrases though - and from your post I suspect you're actually more interested in recognizing [named entities](https://spacy.io/usage/linguistic-features#named-entities). The noun chunker is not a good fit for this.

The [Entity Ruler](https://spacy.io/usage/rule-based-matching#entityruler) relies heavily on user-defined patterns, so if you do something like this:
```
ruler = EntityRuler(nlp)
patterns = [{"label": "GAME", "pattern": [{"LOWER": "call"}, {"LOWER": "of"}, {"LOWER": "duty"}]}]
ruler.add_patterns(patterns)
nlp.add_pipe(ruler)
``` 
it will generate "Call of Duty" as a `GAME` entity. However, I assume that you don't have a pre-defined dictionary of entities you want to recognize?

I think the approach you're looking for is [training a named entity recognizer](https://spacy.io/usage/training#ner) for this challenge, on your specific entity types. You'll need some training data for this, where you manually annotate some example data, then train a neural model to recognize similar entities.

**Comment by kaykhancheckpoint** [comment]
Hey thanks for the reply but actually i was more interested in creating the chunks similar to how EntityRuler is doing it without actually using EntityRuler. As i found EntityRuler was able to create the correct chunk for  "Call of Duty" to be able to recognize it.

Again im more interested in reproducing the method of chunking the text rather than recognizing entities at this point in time.

**Comment by svlandeg** [comment]
Can you paste a minimal code snippet of how you're using the EntityRuler right now? The EntityRuler only recognises the chunk because you feed it the correct pattern, no?

**Comment by svlandeg** [comment]
If you're interested in chunking, I think you probably just want to write a custom function to produce those noun chunks you're interested in - i.e. modifying [the current implementation](https://github.com/explosion/spaCy/blob/master/spacy/lang/en/syntax_iterators.py#L8) as a custom function.

**Comment by kaykhancheckpoint** [comment]
> Can you paste a minimal code snippet of how you're using the EntityRuler right now? The EntityRuler only recognises the chunk because you feed it the correct pattern, no?

Is the entity ruler not chunking the sentence first before recognizing the patterns? 

Or are you saying the EntityRuler is doing pattern matching without chunking?

**Comment by svlandeg** [comment]
That's right, the EntityRuler doesn't rely on any chunking. It's really just a simple pattern-based search mechanism that can be used with any `nlp` pipeline, even when there's no tagger or parser.
